### ISSUE TITLE
Can’t like tweets: Forbidden

### ISSUE THREAD

**Issue by lucaswerkmeister** [IBBoard/cawbird#117]
**Describe the bug**
Since some time this morning (around 11:00 to 13:00 UTC) I’ve been unable to like any tweets in Cawbird: the only error message is “Forbidden”.

**To Reproduce**
1. Go to your timeline
2. Find a tweet you want to like
3. Left click or right click
4. Click the “heart” button
5. Wait a few seconds

**Expected behavior**
The tweet is liked. If right click was used, the tweet goes back to the normal view.

**Screenshots**
![Screenshot from 2020-03-21 14-44-47](https://user-images.githubusercontent.com/2346599/77227708-88026400-6b82-11ea-8817-8e94fc34bc29.png)

**System details:**
 - OS: Arch Linux
 - Version: app version 1.0.4, package version v1.0.4.r59.gfac1d1ff-1
 - Package type: [cawbird-git AUR package](https://aur.archlinux.org/packages/cawbird-git)

**Additional context**
This might just be a transient problem, but if it persists I figured it’s best to report it as soon as possible. I’m pretty sure I managed to like *some* tweets earlier this morning, so if it was a change on Twitter’s side, it must have been very recent.

I can still like tweets in the web interface (and Cawbird will later recognize and display those likes). I can also still retweet tweets in Cawbird, but undoing the retweet is Forbidden, as is undoing a like.

There’s not much more detail in the journal:
```
TweetUtils.vala:130: Exception: Forbidden in TweetUtils.vala:130
TweetUtils.vala:130: Exception: Forbidden in TweetUtils.vala:130
TweetUtils.vala:179: Exception: Forbidden in TweetUtils.vala:179
```

(Those are the error messages for attempting to like a tweet via right click, like it via left click, and un-retweet it, respectively.)

**Comment by IBBoard** [comment]
Looks like I need to improve Cawbird's logging. Twitter gives some good details on [error codes](https://developer.twitter.com/en/docs/basics/response-codes), but Baedert apparently just printed the HTTP response message (because these errors don't normally happen).

**Comment by IBBoard** [comment]
Quickly added some logging.

We're getting "Application cannot perform write actions. Contact Twitter Platform Operations through https://support.twitter.com/forms/platform".

And if I go to the Developer site then I get this:
![Screenshot_2020-03-21 Twitter Developers](https://user-images.githubusercontent.com/388582/77228189-926e2f00-6b7d-11ea-9a06-52ecb126ead9.png)

I guess we're waiting on Twitter's whims as to when they respond to my request through that form.

**Comment by lucaswerkmeister** [comment]
Thanks, let’s see what happens with Twitter… (sending tweets still works, so that “read access only” is a bit curious ^^)

**Comment by IBBoard** [comment]
"Your request has been submitted to Twitter. We are usually able to respond within a few days, but some issues may take longer." - We could be here a while 🙄

I'll try to improve the user feedback in the meantime.

**Comment by IBBoard** [comment]
![screenshot](https://user-images.githubusercontent.com/388582/77228840-fdba0000-6b81-11ea-89ff-88cd07f22d50.png)
Apparently Cawbird may have breached a policy somewhere. But it's not clear where, because they reactivated it within five minutes, didn't say what we did wrong, and then threatened that the next violation will be a permanent ban.

I've asked for clarification. I'm not holding out much hope. Twitter have been historically bad at telling people what the actual problem is.

**Comment by IBBoard** [comment]
Since this did appear to be a Twitter issue, I'm going to close it and deal with better error handling as part of #41.

**Comment by lucaswerkmeister** [comment]
Hm, does liking tweets work for you now? I still get the same error.

**Comment by IBBoard** [comment]
Erm, good point. No, it doesn't. I assumed that when they said they'd fixed it then they'd fixed it. Can't remember whether I actually tested it.

It still says "Restricted to API read access only" in the Developer area, so either they didn't fix it, or they've blocked it again. Great.

**Comment by lucaswerkmeister** [comment]
Workaround: use the [Nextbird API keys](https://github.com/Nextbird/nextbird/commit/c50896ee226a67d54db5b3614b11831cd2321a97) 🤷

```diff
diff --git a/data/uk.co.ibboard.cawbird.gschema.xml b/data/uk.co.ibboard.cawbird.gschema.xml
index 00087767..415c7a76 100644
--- a/data/uk.co.ibboard.cawbird.gschema.xml
+++ b/data/uk.co.ibboard.cawbird.gschema.xml
@@ -88,11 +88,11 @@
       <summary>Size and position of the settings dialog</summary>
     </key>
     <key type="s" name="consumer-key">
-      <default>"Vf9torDW2ZOw32DfhUtI9csL8"</default>
+      <default>"66oowkKJcNxsfaBft7NOYBjub"</default>
       <summary>The application's oauth consumer key.</summary>
     </key>
     <key type="s" name="consumer-secret">
-      <default>"18BEr1mdDH46cJhw5mUMwHe2TiBExOopEDxFbPzfJrlnFuvZJ2"</default>
+      <default>"hfcXJFmqVAGK98wDbeCowQj272Ib18vcslOhvxW17XRLHcLKFJ"</default>
       <summary>The application's oauth consumer secret</summary>
     </key>
     <key type="b" name="round-avatars">
diff --git a/src/CbUserStream.c b/src/CbUserStream.c
index 939f9e08..2c35c589 100644
--- a/src/CbUserStream.c
+++ b/src/CbUserStream.c
@@ -141,16 +141,16 @@ cb_user_stream_init (CbUserStream *self)
 
   if (self->stresstest)
     {
-      self->proxy = oauth_proxy_new ("Vf9torDW2ZOw32DfhUtI9csL8",
-                                     "18BEr1mdDH46cJhw5mUMwHe2TiBExOopEDxFbPzfJrlnFuvZJ2",
+      self->proxy = oauth_proxy_new ("66oowkKJcNxsfaBft7NOYBjub",
+                                     "hfcXJFmqVAGK98wDbeCowQj272Ib18vcslOhvxW17XRLHcLKFJ",
                                      "https://api.twitter.com/",
                                      FALSE);
     }
   else
     {
       /* TODO: We should be getting these from the settings */
-      self->proxy = oauth_proxy_new ("Vf9torDW2ZOw32DfhUtI9csL8",
-                                     "18BEr1mdDH46cJhw5mUMwHe2TiBExOopEDxFbPzfJrlnFuvZJ2",
+      self->proxy = oauth_proxy_new ("66oowkKJcNxsfaBft7NOYBjub",
+                                     "hfcXJFmqVAGK98wDbeCowQj272Ib18vcslOhvxW17XRLHcLKFJ",
                                      "https://api.twitter.com/",
                                      FALSE);
     }
```

Delete account (in-app), relogin, works like a charm. (The app won’t let you log in while the old account for the same name still exists, that’s why the first step is necessary. Also, don’t be alarmed, the app will take a while to start the first time after this, some disk I/O according to `htop`.)

**Comment by IBBoard** [comment]
It is our keys that are blacklisted, so alternate keys will work.

I can't recommend that as an "approved" approach because then it would look like we were working around the ban.

However, I also can't prevent people forking Cawbird, creating their own private developer account and their own keys, and putting the keys into their local version of the app. It is, after all, open source. (But you risk whatever the wrath of Twitter is for circumventing the app's lock-down).

No update from Twitter on what the problem is yet.

**Comment by IBBoard** [comment]
Success!

> Twitter has automated systems that find and disable abusive API keys in bulk. Unfortunately, it looks like your application got caught up in one of these spam groups by mistake. We have reactivated the API key and we apologize for the inconvenience. 
>
> Regards,
> Twitter Platform Operations

I've asked them to confirm a) whether I caused it by spamming a few DMs to myself very quickly so that I had something to test with and b) whether the strike has been entirely removed so that we don't go straight to auto-ban the next time they accidentally flag it!

I may need to be more careful with testing in future.

**Comment by lucaswerkmeister** [comment]
Great, thanks!

If you want, feel free to steal the Nextbird API keys for future tests, if that client gets disabled it’s not a big deal ^^

**Comment by IBBoard** [comment]
😄 Yeah, that's one idea! The problem is that this was more "hmmmm, I need some test data - let's send a load of messages quickly with keyboard shortcuts" rather than 'proper' structure testing where swapping keys isn't that painful.

**Comment by Scotty-Trees** [comment]
Greetings, 

Love you app and thanks for forking it from the original Corebird. With that said, I've run into this issue as well today. Not sure if it's come back or if there's something else going on. I tried liking tweets and I get the "Application cannot perform write actions." everytime I do. I wasn't sure if I need to open a new issue or comment here, but feel free to let me know. For reference I'm running EndeavourOS Gnome 41.1 Cawbird v1.4.2, let me know if you need anymore info or logs from me to help out.

![Screenshot from 2021-11-15 01-53-34](https://user-images.githubusercontent.com/45135187/141736019-e6990933-02aa-4b96-87e5-6aee35c8de48.png)

**Comment by IBBoard** [comment]
Yeah, [we're aware of it](https://twitter.com/CawbirdClient/status/1459255439024173062).

It's most likely a different cause than last time (which was my testing triggering automatic filters) but Twitter Support aren't being very forthcoming with any information. We only found out when people reported issues.

I'll track it via #407 for now.

**Comment by Scotty-Trees** [comment]
@IBBoard No worries, thank you very much for the update and the links! I'll keep track of it as well. Good luck with Twitter and hopefully resolving it at some point in the hopefully not too distant future!
